reana_job_controller/job_monitor.py

```diff
@@ -241,6 +261,87 @@ def watch_jobs(self, job_db, app):
                 time.sleep(120)
 
 
+slurmJobStatus = {
+    'failed': ['BOOT_FAIL', 'CANCELLED', 'DEADLINE', 'FAILED', 'NODE_FAIL',
+               'OUT_OF_MEMORY', 'PREEMPTED', 'TIMEOUT', 'SUSPENDED',
+               'STOPPED'],
+    'succeeded': ['COMPLETED'],
+    'running': ['CONFIGURING', 'COMPLETING', 'RUNNING', 'STAGE_OUT'],
+    'idle': ['PENDING', 'REQUEUE_FED', 'REQUEUE_HOLD', 'RESV_DEL_HOLD',
+             'REQUEUED', 'RESIZING']
+    # 'REVOKED',
+    # 'SIGNALING',
+    # 'SPECIAL_EXIT',
+}
+
+
+@singleton
+class JobMonitorSlurmCERN(JobMonitor):
+    """Slurm jobs monitor CERN."""
+
+    def __init__(self, app=None):
+        """Initialize Slurm job monitor thread."""
+        super(__class__, self).__init__(
+            thread_name='slurm_job_monitor'
+        )
+
+    def format_slurm_job_query(self, backend_job_ids):
+        """Format Slurm job query."""
+        cmd = 'sacct --jobs {} --noheader --allocations --parsable ' \
+              '--format State,JobID'.format(','.join(backend_job_ids))
+        return cmd
+
+    def watch_jobs(self, job_db, app=None):
+        """Use SSH connection to slurm submitnode to monitor jobs.
+
+        :param job_db: Dictionary which contains all running jobs.
+        """
+        slurm_connection = SSHClient(
+            hostname=config.SLURM_HEADNODE_HOSTNAME,
+            port=config.SLURM_HEADNODE_PORT,
+        )
+        statuses_to_skip = ['succeeded', 'failed']
+        while True:
+            logging.debug('Starting a new stream request to watch Jobs')
+            try:
+                slurm_jobs = {}
+                for id, job_dict in job_db.items():
+                    if (not job_db[id]['deleted'] and
+                       job_db[id]['compute_backend'] == 'slurmcern' and
+                       not job_db[id]['status'] in statuses_to_skip):
+                        slurm_jobs[job_dict['backend_job_id']] = id
+                if not slurm_jobs.keys():
+                    logging.error('No slurm jobs')
+                    pass
+                slurm_query_cmd = self.format_slurm_job_query(
+                    slurm_jobs.keys())
+                stdin, stdout, stderr = \
+                    slurm_connection.ssh_client.exec_command(slurm_query_cmd)
+                stdout = stdout.read().decode("utf-8").rstrip().split('\n')
+                for item in stdout:
```

Comment:
Got a traceback around here (but the job goes on); we can perhaps sanitise this part anyway:

```
2019-12-03 14:01:09,552 | werkzeug | Thread-2 | INFO | 127.0.0.1 - - [03/Dec/2019 14:01:09] "POST /jobs HTTP/1.1" 201 -
2019-12-03 14:01:09,561 | werkzeug | Thread-5 | INFO | 127.0.0.1 - - [03/Dec/2019 14:01:09] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
2019-12-03 14:01:09,569 | werkzeug | Thread-6 | INFO | 127.0.0.1 - - [03/Dec/2019 14:01:09] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
2019-12-03 14:01:09,665 | root | slurm_job_monitor | ERROR | Unexpected error: list index out of range
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/reana_job_controller/job_monitor.py", line 323, in watch_jobs
    slurm_job_id = item.split('|')[1]
IndexError: list index out of range
2019-12-03 14:01:12,576 | werkzeug | Thread-7 | INFO | 127.0.0.1 - - [03/Dec/2019 14:01:12] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
2019-12-03 14:01:15,585 | werkzeug | Thread-8 | INFO | 127.0.0.1 - - [03/Dec/2019 14:01:15] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
2019-12-03 14:01:18,595 | werkzeug | Thread-9 | INFO | 127.0.0.1 - - [03/Dec/2019 14:01:18] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
```

Reply:
It's expected, job monitor started iteration while there were no job submitted yet.

Reply:
Also getting "No slurm jobs" later:

```
2019-12-03 14:06:07,668 | werkzeug | Thread-105 | INFO | 127.0.0.1 - - [03/Dec/2019 14:06:07] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
2019-12-03 14:06:10,075 | root | slurm_job_monitor | ERROR | No slurm jobs
2019-12-03 14:06:10,136 | root | slurm_job_monitor | ERROR | Unexpected error: list index out of range
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/reana_job_controller/job_monitor.py", line 323, in watch_jobs
    slurm_job_id = item.split('|')[1]
IndexError: list index out of range
2019-12-03 14:06:10,686 | werkzeug | Thread-111 | INFO | 127.0.0.1 - - [03/Dec/2019 14:06:10] "GET /jobs/3d61f10e-a8a4-4d1a-9983-8d28a5de16e3 HTTP/1.1" 200 -
2019-12-03 14:06:14,540 | werkzeug | Thread-112 | INFO | 127.0.0.1 - - [03/Dec/2019 14:06:14] "POST /jobs HTTP/1.1" 201 -
2019-12-03 14:06:14,564 | werkzeug | Thread-113 | INFO | 127.0.0.1 - - [03/Dec/2019 14:06:14] "GET /jobs/38d1a1e1-615f-4aea-9afe-02ebcec02159 HTTP/1.1" 200 -
2019-12-03 14:06:14,579 | werkzeug | Thread-114 | INFO | 127.0.0.1 - - [03/Dec/2019 14:06:14] "GET /jobs/38d1a1e1-615f-4aea-9afe-02ebcec02159 HTTP/1.1" 200 -
```

Reply:
> It's expected, job monitor started iteration while there were no job submitted yet.

You can catch this, if there is no output to parse, it is not necessary to try to extract IDs... The logs will be cleaner that way.